### PR TITLE
Add measurements and timers to the Statistics base class

### PR DESCRIPTION
**Pull request by haved**
While modifying Andersen, I tried to evolve the `Andersen::Statistics` class with the changes, and was annoyed by two things:

 - Modifying the set of tracked measurements required 4 changes to the code (ctor init, field, assigning the value and serializing the value)
 - There was no nice way of making certain fields optional
 - Also, tests don't have access to statistics

In this PR is a suggestion for a modified base class for `Statistics`. I also modified `Andersen::Statistics` to make use of it. This PR should of course also have tests added, but I'm putting it up now to get feedback.

This PR should correspond to #361

### CODE REVIEW

File: jlm/util/Statistics.hpp [hunk 1]
```diff
@@ -64,8 +75,164 @@ class Statistics
...
+  // Mutable version of @see GetTimer
+  [[nodiscard]] util::timer &
+  GetTimer(const std::string & name);
+
+  /**
+   * Removes the timer with the given \p name, if it exists.
+   * @return true if the timer existed and was removed, false otherwise
+   */
+  bool
+  RemoveTimer(const std::string & name);
```
Comment (phate) [inline jlm/util/Statistics.hpp:228]: This interface feels a little bit inconsequent to me:

1. I would remove `RemoveTimer()` from it. Do we have a use case for this method?
2. You are offering several methods reflecting the operations on a timer, but at the same time also return the underlying timer for each of those methods, which could be used to perform the same operations.

Alternative interface suggestion:

```
util::timer &
AddTimer(std::string name);

[[nodiscard]] util::timer &
GetTimer(const std::string & name);
```

And then rely on the timer interface for performing the necessary operations. In addition, I would delete the copy constructor and copy assignment operator, as it would be way too easy to do this mistake:

`auto timer  = GetTimer("myTimer")` instead of `auto & timer = GetTimer("myTimer")`.

File: jlm/util/time.hpp [hunk 2]
```diff
@@ -15,29 +15,71 @@ class timer final
 {
 public:
   constexpr timer()
+      : CountedNanoseconds_(0),
```
Comment (phate) [inline jlm/util/time.hpp:18]: How about `ElapsedTimeInNanoseconds`?

File: jlm/util/time.hpp [hunk 3]
```diff
@@ -15,29 +15,71 @@ class timer final
...
+        Running_(false)
```
Comment (phate) [inline jlm/util/time.hpp:19]: How about `IsRunning_`?

File: jlm/llvm/opt/alias-analyses/Andersen.cpp [hunk 4]
```diff
@@ -34,113 +34,104 @@ class Andersen::Statistics final : public jlm::util::Statistics
...
 
   void
   StartAndersenStatistics(const jlm::rvsdg::graph & graph) noexcept
   {
-    NumRvsdgNodes_ = jlm::rvsdg::nnodes(graph.root());
-    AnalysisTimer_.start();
-    SetAndConstraintBuildingTimer_.start();
+    AddUInteger("#RvsdgNodes", jlm::rvsdg::nnodes(graph.root()));
+    StartTimer("AnalysisTimer");
+    StartTimer("SetAndConstraintBuildingTimer");
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:45]: Could factor out these magic strings into variables? In particular, those strings that are used multiple times, such as "SetAndConstraintBuildingTimer"?

File: jlm/llvm/opt/alias-analyses/Andersen.cpp [hunk 5]
```diff
@@ -150,77 +141,7 @@ class Andersen::Statistics final : public jlm::util::Statistics
...
-        " ",
-        "ConstraintSolvingTime[ns]:",
-        ConstraintSolvingTimer_.ns(),
-        " ",
-        "PointsToGraphConstructionTime[ns]:",
-        PointsToGraphConstructionTimer_.ns(),
-        " ",
-        "PointsToGraphConstructionExternalToEscapedTime[ns]:",
-        PointsToGraphConstructionExternalToEscapedTimer_.ns());
+        SerializeMeasurements());
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:144]: Could you add another `SerializeMeasurements()` method that prefixes the measurements with the Statistics::Id followed by the source file. Ultimately, this is what we want (and after we changed everything to use that one, we remove the ToString() method). However, I understand that the current method is necessary in order to not have the changes all at once all over the place. I can take care in the end to refactor all the other statistics.

File: jlm/util/Statistics.cpp [hunk 6]
```diff
@@ -1,15 +1,230 @@
...
 
+#include <jlm/util/strfmt.hpp>
+
 namespace jlm::util
 {
 
 Statistics::~Statistics() = default;
 
+std::string
+Statistics::SerializeMeasurements() const
```
Comment (phate) [inline jlm/util/Statistics.cpp:17]: This method does more than the name suggests (it also serializes the timers). How about just naming it `Serialize()`?

**Review by phate — CHANGES_REQUESTED** [review]
Thanks for the initiative. I really appreciate that you try to improve the infrastructure instead of just living with it!

Could we split this in 3 individual PRs:
1. A PR for the timer. I just had minor suggestions there. I would also add some unit tests (I know there are none for it currently).
2. A PR for the changes in statistics. Overall, I am very positive to the change. It addresses the concerns I mentioned [here](https://github.com/phate/jlm/issues/361). We need to hammer out the details though. I would also add some unit tests on this PR.
3. A PR where you use the new interface in Andersen. Since the changes allow now to easily write unit tests, I would also add one here.

File: jlm/llvm/opt/alias-analyses/Andersen.cpp (hunk 5 above)
Comment (haved) [reply jlm/llvm/opt/alias-analyses/Andersen.cpp:144]: I tried doing something like that at first (giving `ToString` a default implementation) but encountered two issues:

 - where can I find the ID as a string? I'm guessing it eventually should correspond to the ID used on the command line, but right now there is no source of truth location for translating between the enum and a string.
 - not all Statistics have a field for the file name. The `RegionAwareMemoryNodeProvider`'s statistics don't include it in it's output. Should all Statistics have the fileName included? If so, we can maybe move the field to the superclass and add it to the ctor?

File: jlm/llvm/opt/alias-analyses/Andersen.cpp (hunk 4 above)
Comment (haved) [reply jlm/llvm/opt/alias-analyses/Andersen.cpp:45]: I'm a bit hesitant to put all the strings into variables. Some of the purpose of this PR was to reduce the number of separate lines of code that need to be modified whenever a statistic is added, removed or changed.

If we were to add variables for these strings, they would either only be accessible within the `.cpp` file, where most of them only see one use, or they would have to be in the `.hpp` file, which makes it somewhat easier for tests to retrieve measurements, but now requires you to modify two files whenever a statistic is added/removed/modified.

The only place I can see where having access to the strings is useful is when retrieving statistics in tests. By using variables instead of magic strings, IDEs can correctly perform renaming across files, which is nice. However, a test failure makes itself apparent "almost" as quickly as a compilation error, so I don't think it is worth it to place all string literals in variables in header files for only tests' sake.

For the timers, I agree, it is a bit annoying having to reference them by name twice. It will assert false if there is a mismatch, but variables would make it clearer. The question is if the variable should be a reference to the timer object, or a string.

**Comment by haved** [comment]
@phate I pushed my changes to resolve as many of your comments as I could, but this PR will still be turned into two separate PRs. I just want your feedback on my response to one of your comments, before I make the Statistics class PR.

File: jlm/llvm/opt/alias-analyses/Andersen.cpp (hunk 4 above)
Comment (phate) [reply jlm/llvm/opt/alias-analyses/Andersen.cpp]: I should have been a little bit more elaborate in my original comment. Sorry! Why I mentioned the magical strings was that you are requesting timers multiple times (at least twice, once for start and once for stop), and in both cases you are using a string. Thus, instead of writing something like this:

```
class Statistics
{
public:
  void
  StartMySpecialStatistics()
  {
    StartTimer("MySpecialStatisticsTimer");
    ...
  }

  void
  StopMySpecialStatistics()
  {
     StopTimer("MySpecialStatisticsTimr"); // typo here
  }
}
```

I have something like this in mind:
```
class Statistics
{
private:
  const char * MySpecialStatisticsTimerTag_ = "MySpecialStatisticsTimer";
public:
  void
  StartMySpecialStatistics()
  {
    StartTimer(MySpecialStatisticsTimerTag_);
    ...
  }

  void
  StopMySpecialStatistics()
  {
     StopTimer(MySpecialStatisticsTimerTag_);
  }
}
```

This was only meant for:
1. The statistics class internally and was not supposed to "leak" into tests/outside.
2. Only for instances were you have to use them multiple times (to avoid typos, and make your life easier in terms of renaming by being able to utilize the IDE)
3. The idea was to avoid typos and potentially reference a wrong timer, but looking at the code you would throw an exception if one would have a typo. So this argument does not really count any longer (my bad, should have taken a closer look at the code).

I do not really see that the "additional number of lines" argument counts here as most of the renaming with those tags would be done by the IDE in the end. Personally, I would still try to use such tags for entities that I have to reference at least multiple times, but this is just personal preference.

Due to the exception, the original argument is void. Leave it as it is if you like. No strong opinion about this.

File: jlm/llvm/opt/alias-analyses/Andersen.cpp (hunk 5 above)
Comment (phate) [reply jlm/llvm/opt/alias-analyses/Andersen.cpp]: 1. As far as I know, there is currently no method for providing a string representation of a statistics ID. C++ does not come equipped with it, and I have not implemented one so far. You would need to implement it. We have methods for translating the statistics IDs to command line flags though. See [here](https://github.com/phate/jlm/blob/94ba20fd93ace0da2a0b52df6891cc3b46051bb4/jlm/tooling/CommandLine.cpp#L180C1-L269C2). I am a little bit reluctant of tying the internal string representation of an statistics id to the command line flag though. It would mean that by changing the name of an id (and its string representation), one would all in a sudden also change the command line flag, which is a public interface. I think they should be separate concerns.
2. Yes, all statistics should have the filename included. It makes no sense to leave this (important) piece of information out. Moving it to the superclass and adding it to the ctor would be the right thing to do.

**Comment by phate** [comment]
@haved We can close this PR, right? It was replaced by the three individual PRs.

**Comment by haved** [comment]
@phate Yes, closing it now
